[Mobile] - Adds useEditorWrapperStyles hook

geriux opened the pull request:
Closes https://github.com/wordpress-mobile/gutenberg-mobile/issues/5614

## What?
This PR introduces a new hook `useEditorWrapperStyles` which will replace the usage of the [ReadableContentView](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/components/src/mobile/readable-content-view/index.native.js) component in a separate PR. It will also handle the margins and max-width that will be used for the block wrappers within the `BlockList`.

## Why?
Currently, the [ReadableContentView](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/components/src/mobile/readable-content-view/index.native.js) component uses listeners to detect window dimension changes to then partly handle the alignment of the content depending on the `align` prop. This component is used in different places like the [EditorHeader](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/edit-post/src/components/visual-editor/header.native.js), [EmptyListComponent](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/block-editor/src/components/block-list/index.native.js#L444), and [BlockListItem](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/block-editor/src/components/block-list/block-list-item.native.js#L122).

Having this in the `BlockListItem` for example when having different blocks that use InnerBlocks means that there will be quite a lot of listeners added when is not needed.

Nesting too many `View` components is also a problem as we've reached the maximum in the past so we want to avoid that too.

The [BlockListItem](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/block-editor/src/components/block-list/block-list-item.native.js#L122) component also handles margin and style data depending on the screen and the alignment so this hook will take care of that too.

## How?

The hook will use there functions:

- `getWideWidthStyles`-> It will handle the styles logic for `wide` alignments.
- `getFullWidthStyles`-> It will handle the styles logic for `full` alignments.
- `getBlockMargin` -> It will handle the horizontal margins for a block taking into account its parents.

### useEditorWrapperStyles

Will return the `wrapperStyles` and the `margin`.

**Usage examples:**

For the [EditorHeader](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/edit-post/src/components/visual-editor/header.native.js) component:

```diff
 	} ) {
+		const [ wrapperStyles ] = useEditorWrapperStyles();
 		const blockHolderFocusedStyle = getStylesFromColorScheme(
 			styles.blockHolderFocused,
 			styles.blockHolderFocusedDark
 		);
+
 		return (
-			<ReadableContentView>
+			<View style={ wrapperStyles }>
 				<PostTitle
```

For the [BlockListItem](https://github.com/WordPress/gutenberg/blob/9091d91143160e171860201b799d056c9496b288/packages/block-editor/src/components/block-list/block-list-item.native.js#L122) component

```diff
+	const [ wrapperStyles, margin ] = useEditorWrapperStyles( {
+		align: blockAlignment,
+		blockName,
+		blockWidth,
+		contentResizeMode,
+		hasParents,
+		parentBlockAlignment,
+		parentBlockName,
+		parentWidth,
+	} );
+
+	return (
+		<View
+			style={ wrapperStyles }
+			pointerEvents={ isReadOnly ? 'box-only' : 'auto' }
+		>
+			{ shouldShowInsertionPointBefore && <BlockInsertionPoint /> }
+			<BlockListBlock
+				key={ clientId }
+				showTitle={ false }
+				clientId={ clientId }
+				parentWidth={ parentWidth }
+				{ ...restProps }
+				marginHorizontal={ margin }
+				blockWidth={ blockWidth }
+			/>
+			{ ! shouldShowInnerBlockAppender() &&
+				shouldShowInsertionPointAfter && <BlockInsertionPoint /> }
+		</View>
+	);
```

## Testing Instructions
CI checks should pass.

An integration PR will follow up this one.

### Testing Instructions for Keyboard
N/A

## Screenshots or screencast <!-- if applicable -->
N/A

Commit:
Mobile - Adds initial useEditorWrapperStyles hook

Commit:
Mobile - Update useEditorWrapperStyles hook to take into account custom marginHorizontal values

dcalhoun commented on packages/block-editor/src/hooks/test/use-editor-wrapper-styles.native.js:
```diff
@@ -0,0 +1,475 @@
+	useEditorWrapperStyles,
+} from '../use-editor-wrapper-styles';
+
+jest.mock( '../use-editor-wrapper-styles.scss', () => ( {
```
I interpret this to be an alternative to [`styleMock.js`](https://github.com/WordPress/gutenberg/blob/bd432f149026c35d07edd05c7de5af8e90e16ff0/test/native/__mocks__/styleMock.js#L1). Colocating this mock makes a lot more sense to me. 👍🏻

geriux replied:
Exactly, because our current [setup overrides](https://github.com/WordPress/gutenberg/blob/trunk/test/native/jest.config.js#L47) all `.scss` files with the `styleMock` file. Maybe in the future, we should keep mocking styles within the test itself.

dcalhoun commented on packages/block-editor/src/hooks/use-editor-wrapper-styles.native.js:
```diff
@@ -0,0 +1,250 @@
+ * @param {number}  width                The width of the screen.
+ * @return {number} The calculated block margin.
+ */
+export function getBlockMargin(
+	align,
+	blockName,
+	blockWidth,
+	hasParents,
+	marginHorizontal,
+	parentBlockAlignment,
+	parentBlockName,
+	parentWidth,
+	width
+) {
```
IMO functions with more than 1-2 parameters are quite cumbersome to both use and comprehend existing usage, primarily the latter. Would you be open to using named arguments in the form of an object for all or some of these arguments? 

This thought applies to other functions in this file as well.

geriux replied:
> Would you be open to using named arguments in the form of an object for all or some of these arguments?
> This thought applies to other functions in this file as well.

Definitely! That's a great suggestion, I'll make the changes 👍

geriux replied:
This was updated in https://github.com/WordPress/gutenberg/pull/49616/commits/33d4472ac8d2bd57938c62aca84d2ec4375a5d11

dcalhoun commented on packages/block-editor/src/hooks/use-editor-wrapper-styles.native.js:
```diff
@@ -0,0 +1,250 @@
+		! isContainerRelated( parentBlockName ) &&
+		isContainerRelated( blockName )
+	) {
+		fullWidthStyles.paddingHorizontal =
+			styles[
+				'block-editor-hooks__use-editor-wrapper-styles-block'
+			].marginLeft;
```
It appears this selector (and many others referenced in this file) only contain a single attribute. Do we see value in using single-attribute styles (Sass) rather than holding the values in the JavaScript file?

geriux replied:
>  Do we see value in using single-attribute styles (Sass) rather than holding the values in the JavaScript file?

Not really, I'd say only for `block-editor-hooks__use-editor-wrapper-styles-block` because it uses the value of `$block-edge-to-content` but we also have that value `16` in JavaScript files.

If we move it to the JavaScript side, we could have simpler constant names instead of these long 😅 

I'll move them and keep only the wrapper styles in the styles file.

geriux replied:
Updated in https://github.com/WordPress/gutenberg/pull/49616/commits/696d5a97dabea9f96168ccdf80af6b2c8491df3b#diff-48fae6ad7c03074aba0381f663a92c2882e22ef6504fe8ae49a4b49fa5f0161f

dcalhoun commented on packages/block-editor/src/hooks/use-editor-wrapper-styles.native.js:
```diff
@@ -0,0 +1,250 @@
+
+	const blockMargin = useMemo(
+		() =>
+			getBlockMargin(
```
In addition to the internal usage of `getBlockMargin` here, I note it is exported as well. Is the intent that this is a public method as well? If not, we should likely avoid exporting it. 

This thought also applies to other functions in this file.

dcalhoun replied:
> If not, we should likely avoid exporting it.

Related to this point, I'm less inclined to write and maintain automated tests for any "private" methods. Testing the "public" `useEditorWrapperStyles` should be sufficient IMO.

geriux replied:
Yeah, I didn't think about that and maintaining them until I made new changes and I had to go back and update them again 🙃 . I will keep only the tests for `useEditorWrapperStyles` and since the others will be removed, we won't need to export the private functions 👍

geriux replied:
Both suggestions updated in https://github.com/WordPress/gutenberg/pull/49616/commits/cc984f2a36e087ffd8c8aba9f6577839b2964b40

dcalhoun commented on packages/block-editor/src/hooks/use-editor-wrapper-styles.native.scss:
```diff
@@ -0,0 +1,31 @@
+.block-editor-hooks__use-editor-wrapper-styles {
```
I think we could drop `block-editor-hooks` from the selector names. I'm not sure the context it adds is all that meaningful and it significantly elongates the selectors. WDYT?

geriux replied:
I agree, it does make the names to be too long. I'll make the change 👍

geriux replied:
Updated in https://github.com/WordPress/gutenberg/pull/49616/commits/696d5a97dabea9f96168ccdf80af6b2c8491df3b#diff-5fe52ed070a8b8421565978bb3b671eb1b0d89d746db2b63a7203b52c937e0a7

dcalhoun reviewed (commented):
Thanks for providing such great context in the PR description. Very helpful.

Commit:
Mobile - useEditorWrapperStyles hook - Remove helper tests and exports

Commit:
Mobile - useEditorWrapperStyles - Move max width values from the SCSS styles to the JavaScript side and update tests

Commit:
Mobile - useEditorWrapperStyles - Update helpers to use an options object param instead of first level params

geriux commented:
Hey @dcalhoun 👋 I've addressed the first round of feedback, let me know what you think! I also updated the feature `rnmobile/block-list-refactor` branch and this branch with `trunk`

dcalhoun commented on packages/block-editor/src/hooks/use-editor-wrapper-styles.native.js:
```diff
@@ -0,0 +1,249 @@
+		// For these cases, no width constraints should be added.
+		if ( stretchStyle ) {
+			canvasStyles = stretchStyle;
+		}
```
I wonder if it might be clearer to combine these conditionals into one, rather than conditionally mutating. WDYT? 

```js
let canvasStyles;
if (contentResizeMode === "stretch") {
  // For these cases, no width constraints should be added.
  canvasStyles = { flex: 1 };
} else if (reversed) {
  canvasStyles = styles["use-editor-wrapper-styles--reversed"];
} else {
  canvasStyles = styles["use-editor-wrapper-styles"];
}
```

geriux replied:
I like that suggestion 😃 , I'll update it before merging. Thanks!

geriux replied:
Updated in https://github.com/WordPress/gutenberg/pull/49616/commits/1d281e0d09e73492a4416f121164d38c9fe385e4

dcalhoun reviewed (approved):
The latest changes look great. My one comment is a suggestion, not a request.

Commit:
Mobile - useEditorWrapperStyles - Update wrapperStyles logic to unify canvasStyles logic

geriux commented:
> The latest changes look great. My one comment is a suggestion, not a request.

Thank you for the review and feedback!
